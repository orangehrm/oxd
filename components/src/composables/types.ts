import {InjectionKey, Ref} from 'vue';

export type ErrorField = {
  cid: string;
  errors: string[];
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Rule = (value?: any) => string | boolean;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ModelValue = Ref<any>;

export type ErrorBag = Array<ErrorField>;
export type Fieldset = Array<FormField>;

export interface FormField {
  cid: Ref<string>;
  label: Ref<string>;
  dirty: Ref<boolean>;
  touched: Ref<boolean>;
  processing: Ref<boolean>;
  validate: () => Promise<ErrorField>;
  reset: () => void;
}

export interface FormAPI {
  searchErrors: (id: string) => ErrorField[];
  purgeErrors: (id: string) => void;
  addError: (errorField: ErrorField) => void;
  registerField: (field: FormField) => void;
  unregisterField: (field: FormField) => void;
}

export const formKey: InjectionKey<FormAPI> = Symbol('form');
