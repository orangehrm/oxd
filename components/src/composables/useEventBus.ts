interface EventQueue {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: Array<(data: any) => void>;
}

class EventBus {
  private queue: EventQueue;
  constructor() {
    this.queue = {};
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  $on(name: string, callback: (data: any) => void) {
    this.queue[name] = this.queue[name] || [];
    this.queue[name].push(callback);
  }

  $off(name: string, callback: () => void) {
    if (this.queue[name]) {
      for (let i = 0; i < this.queue[name].length; i++) {
        if (this.queue[name][i] === callback) {
          this.queue[name].splice(i, 1);
          break;
        }
      }
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  $emit(name: string, data: any) {
    if (this.queue[name]) {
      this.queue[name].forEach(function(callback) {
        callback(data);
      });
    }
  }
}

export default function useEventBus() {
  return new EventBus();
}
