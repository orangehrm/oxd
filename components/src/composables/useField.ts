/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {ref, Ref, watch, computed, onBeforeUnmount, WatchStopHandle} from 'vue';
import {nanoid} from 'nanoid';
import {injectStrict} from '../utils/injectable';
import {ErrorField, FormAPI, formKey, rule} from './types';

export default function useField(fieldContext: {
  fieldLabel: string;
  rules: rule[];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  modelValue: Ref<any>;
  onReset: () => Promise<void>;
  disabled: Ref<boolean>;
}) {
  const form = injectStrict<FormAPI>(formKey);
  const cid = ref<string>(nanoid());
  const label = ref<string>(fieldContext.fieldLabel);
  const dirty = ref<boolean>(false);
  const touched = ref<boolean>(false);
  const processing = ref<boolean>(false);
  let watchHandler: WatchStopHandle;

  const validate = () => {
    if (fieldContext.disabled.value) {
      return Promise.resolve({cid: cid.value, errors: []});
    }

    processing.value = true;
    const allValidations = Promise.all(
      fieldContext.rules.map(func => {
        return new Promise<boolean>((resolve, reject) => {
          Promise.resolve(func(fieldContext.modelValue.value)).then(valid => {
            if (valid === true) {
              resolve(valid);
            } else if (typeof valid === 'string') {
              reject(valid);
            } else {
              reject(
                new Error(
                  `Rules should return a string or true, received '${typeof valid}'`,
                ),
              );
            }
          });
        });
      }),
    );

    return new Promise<ErrorField>((resolve, reject) => {
      allValidations
        .then(() => {
          resolve({
            cid: cid.value,
            errors: [],
          });
        })
        .catch(error => {
          if (typeof error === 'string') {
            resolve({
              cid: cid.value,
              errors: [error],
            });
          } else {
            reject(error);
          }
        })
        .finally(() => {
          processing.value = false;
        });
    });
  };

  const startWatcher = () => {
    watchHandler = watch(fieldContext.modelValue, () => {
      validate().then(result => {
        form.addError(result);
      });
    });
  };

  startWatcher();

  const reset = () => {
    dirty.value = false;
    touched.value = false;
    processing.value = false;
    watchHandler(); // stop the validation watcher
    fieldContext.onReset().then(() => startWatcher());
  };

  form.registerField({cid, label, dirty, touched, processing, validate, reset});

  onBeforeUnmount(() => {
    form.unregisterField({
      cid,
      label,
      dirty,
      touched,
      processing,
      validate,
      reset,
    });
  });

  const hasError = computed(() => {
    return form.searchErrors(cid.value).length > 0;
  });

  const message = computed(() => {
    return hasError.value ? form.searchErrors(cid.value)[0].errors[0] : null;
  });

  return {
    form,
    reset,
    validate,
    hasError,
    message,
  };
}
