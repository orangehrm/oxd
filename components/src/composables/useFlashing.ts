import {ShallowReactive, ref, watch} from 'vue';

export default function useFlashing(
  props: ShallowReactive<{
    order: object;
    items: object[];
    loading: boolean;
  }>,
) {
  let oldValue: object[] = [];
  const flashIndexes = ref(Array(props.items.length).fill(false));

  // Reset cached values if loading or sorting
  watch([() => props.loading, () => props.order], () => (oldValue = []), {
    flush: 'pre',
  });

  watch(
    () => props.items,
    newValue => {
      flashIndexes.value = Array(newValue.length).fill(false);
      for (let i = 0; i < flashIndexes.value.length; i++) {
        if (oldValue.length === 0) break;
        const isNew =
          oldValue.findIndex(item => {
            const oldVal = JSON.stringify(Object.entries(item).sort());
            const newVal = JSON.stringify(Object.entries(newValue[i]).sort());
            return oldVal === newVal;
          }) === -1;
        if (isNew) flashIndexes.value[i] = true;
      }
      oldValue = JSON.parse(JSON.stringify(newValue));
    },
    {deep: true, flush: 'post'},
  );

  return flashIndexes;
}
