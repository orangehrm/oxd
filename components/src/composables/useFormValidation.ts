import {provide, reactive, toRefs, unref} from 'vue';
import {
  ErrorBag,
  ErrorField,
  Fieldset,
  FormAPI,
  FormField,
  formKey,
} from './types';

interface State {
  fieldset: Fieldset;
  errorbag: ErrorBag;
}

export default function useFormValidation() {
  const formState: State = reactive({
    fieldset: [],
    errorbag: [],
  });

  const searchErrors = (id: string) => {
    return formState.errorbag.filter(item => {
      return item.cid === id;
    });
  };

  const purgeErrors = (id: string) => {
    formState.errorbag = formState.errorbag.filter(item => item.cid !== id);
  };

  const addError = (errorField: ErrorField) => {
    purgeErrors(unref(errorField.cid));
    if (errorField.errors.length > 0) formState.errorbag.push(errorField);
  };

  const registerField = (field: FormField) => {
    const i = formState.fieldset.findIndex(item => {
      return unref(item.cid) === unref(field.cid);
    });
    if (i === -1) {
      formState.fieldset.push(field);
    }
  };

  const unregisterField = (field: FormField) => {
    purgeErrors(unref(field.cid));

    const i = formState.fieldset.findIndex(item => {
      return unref(item.cid) === unref(field.cid);
    });
    if (i > -1) {
      formState.fieldset.splice(i, 1);
    }
  };

  const validate = () => {
    return Promise.all(
      formState.fieldset.map(field => {
        if (!field.dirty.value) {
          field.setDirty(true);
          field.startWatcher();
        }
        return field.validate();
      }),
    )
      .then(results => {
        formState.errorbag = [];
        results.map(errorField => {
          addError(errorField);
        });
      })
      .catch(reject => {
        // eslint-disable-next-line no-console
        console.error(reject);
      });
  };

  const reset = () => {
    formState.fieldset.map(field => field.reset());
    formState.errorbag = [];
  };

  provide(formKey, {
    addError,
    purgeErrors,
    searchErrors,
    registerField,
    unregisterField,
  } as FormAPI);

  return {
    ...toRefs(formState),
    reset,
    validate,
    purgeErrors,
  };
}
