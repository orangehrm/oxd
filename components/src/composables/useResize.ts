/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {nextTick, onBeforeUnmount, onMounted, reactive, Ref, toRefs} from 'vue';

export interface State {
  width: number;
  height: number;
}

export default function useResize(
  domElem: Ref<HTMLElement | null | undefined>,
) {
  const state: State = reactive({
    width: 0,
    height: 0,
  });

  const setElementSize = () => {
    state.width = domElem.value?.clientWidth || 0;
    state.height = domElem.value?.clientHeight || 0;
  };

  onMounted(async () => {
    await nextTick();
    setElementSize();
    window.addEventListener('resize', setElementSize);
  });

  onBeforeUnmount(() => {
    window.removeEventListener('resize', setElementSize);
  });

  return {
    ...toRefs(state),
  };
}
