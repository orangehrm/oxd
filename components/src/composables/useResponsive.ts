import {nextTick, onBeforeUnmount, onMounted, reactive, toRefs} from 'vue';

interface Breakpoint {
  [screenType: string]: {
    min: number;
    max: number;
  };
}

export interface State {
  windowWidth: number;
  windowHeight: number;
  screenType: string;
}

export const breakpoints: Breakpoint = {
  xs: {
    min: 0,
    max: 599,
  },
  sm: {
    min: 600,
    max: 799,
  },
  md: {
    min: 800,
    max: 999,
  },
  lg: {
    min: 1000,
    max: 1199,
  },
  xl: {
    min: 1200,
    max: 4000,
  },
};

export default function useResponsive() {
  const state: State = reactive({
    windowWidth: 0,
    windowHeight: 0,
    screenType: '',
  });

  const setWindowSize = () => {
    const width = document.documentElement.clientWidth;
    const height = document.documentElement.clientHeight;
    state.windowWidth = width;
    state.windowHeight = height;
    for (const screenType in breakpoints) {
      if (
        breakpoints[screenType].min <= width &&
        breakpoints[screenType].max >= width
      ) {
        state.screenType = screenType;
        break;
      }
    }
  };

  onMounted(async () => {
    await nextTick();
    setWindowSize();
    window.addEventListener('resize', setWindowSize);
  });

  onBeforeUnmount(() => {
    window.removeEventListener('resize', setWindowSize);
  });

  return state;
}
