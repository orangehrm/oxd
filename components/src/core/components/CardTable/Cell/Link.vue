<template>
  <oxd-table-cell-default>
    <template #body="{row, cell, header}">
      <a
        :href="row.link || '#'"
        :target="header.target || '_self'"
        class="oxd-table-cell-link "
      >
        {{ cell }}
      </a>
    </template>
  </oxd-table-cell-default>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import DefaultCell from './Default.vue';

export default defineComponent({
  components: {
    'oxd-table-cell-default': DefaultCell,
  },
});
</script>

<style src="./link.scss" lang="scss" scoped></style>
