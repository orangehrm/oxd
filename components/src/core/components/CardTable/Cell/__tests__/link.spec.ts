import {mount} from '@vue/test-utils';
import Link from '@orangehrm/oxd/core/components/CardTable/Cell/Link.vue';
import {DEVICE_LG} from '@orangehrm/oxd/composables/useResponsive';

const GLOBAL = {
  provide: {
    tableProps() {
      return {
        items: [],
      };
    },
    screenState() {
      return {
        screenType: DEVICE_LG,
      };
    },
  },
};

describe('CardTable > Cell > Link.vue', () => {
  it('should renders OXD CardTable > Cell > Link Cell', () => {
    const wrapper = mount(Link, {
      global: GLOBAL,
      props: {item: 'This is a link'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should renders OXD CardTable > Cell > Cell with custom link', () => {
    const wrapper = mount(Link, {
      global: GLOBAL,
      props: {
        header: {
          name: 'url',
          target: '_blank',
        },
        rowItem: {
          link: 'https://orangehrm.com',
        },
        item: 'This is a link',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
