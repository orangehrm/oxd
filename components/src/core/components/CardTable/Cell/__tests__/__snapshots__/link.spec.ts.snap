// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTable > Cell > Link.vue should renders OXD CardTable > Cell > Cell with custom link 1`] = `
<div class="oxd-table-card-cell">
  <div class="header"></div><a href="https://orangehrm.com" target="_blank" class="oxd-table-cell-link">This is a link</a>
  <!--v-if-->
</div>
`;

exports[`CardTable > Cell > Link.vue should renders OXD CardTable > Cell > Link Cell 1`] = `
<div class="oxd-table-card-cell">
  <div class="header"></div><a href="#" target="_self" class="oxd-table-cell-link">This is a link</a>
  <!--v-if-->
</div>
`;
